### PR TITLE
[IIP-13] contract staking struct definition (1/3 split by 3853)

### PR DESCRIPTION
**Pull request by envestcc**
# Description
For #3853 is too large to review, it has been splited into three small pr. This is the first one.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/contractstaking/bucket_type.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	if !ok {
+		return errors.New("failed to parse amount")
+	}
+	bt.Duration = p.Duration
+	bt.ActivatedAt = p.ActivatedAt
+	return nil
+}
+
+// Serialize serializes the bucket type
+func (bt *BucketType) Serialize() []byte {
```
Comment (Liuhaai) [inline action/protocol/staking/contractstaking/bucket_type.go:47]: public funs are laid above private funcs

File: action/protocol/staking/contractstaking/bucket.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	Index                     uint64
+	Candidate                 address.Address
+	Owner                     address.Address
+	StakedAmount              *big.Int
+	StakedDurationBlockNumber uint64
+	CreateBlockHeight         uint64
+	StakeStartBlockHeight     uint64
+	UnstakeStartBlockHeight   uint64
+	AutoStake                 bool
+	ContractAddress           string
```
Comment (CoderZhi) [inline action/protocol/staking/contractstaking/bucket.go:25]: comments

File: action/protocol/staking/contractstaking/bucket_info.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	bi.UnstakedAt = p.UnstakedAt
+	bi.Delegate, err = address.FromString(p.Delegate)
+	if err != nil {
+		return err
+	}
+	bi.Owner, err = address.FromString(p.Owner)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (CoderZhi) [inline action/protocol/staking/contractstaking/bucket_info.go:68]:
```
    delegate, err = address.FromString(p.Delegate)
    if err != nil {
        return err
    }
    owner, err := address.FromString(p.Owner)
    if err != nil {
        return err
    }
    bi.TypeIndex = p.TypeIndex
    bi.CreatedAt = p.CreatedAt
    bi.UnlockedAt = p.UnlockedAt
    bi.UnstakedAt = p.UnstakedAt
    bi.Delegate = delegate
    bi.Owner = owner
    return nil
```

File: action/protocol/staking/contractstaking/bucket_type.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	return &contractstakingpb.BucketType{
+		Amount:      bt.Amount.String(),
+		Duration:    bt.Duration,
+		ActivatedAt: bt.ActivatedAt,
+	}
+}
+
+func (bt *BucketType) loadProto(p *contractstakingpb.BucketType) error {
+	var ok bool
+	bt.Amount, ok = big.NewInt(0).SetString(p.Amount, 10)
```
Comment (CoderZhi) [inline action/protocol/staking/contractstaking/bucket_type.go:37]: same here

**Review by CoderZhi — APPROVED** [review]

File: action/protocol/staking/contractstaking/bucket.go (hunk 2 above)
Comment (dustinxie) [inline action/protocol/staking/contractstaking/bucket.go:25]: nit: is this necessary?
will there be another LSD staking contract in the future?

File: action/protocol/staking/contractstaking/bucket.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+}
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/bucket.go:26]: no `Serialize` and `Deserialize` code here?

**Review by dustinxie — APPROVED** [review]

File: action/protocol/staking/contractstaking/bucket.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/bucket.go:25]: yeah

File: action/protocol/staking/contractstaking/bucket.go (hunk 5 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/bucket.go:26]: useless for `Bucket`, which is only in memory

File: action/protocol/staking/contractstaking/bucket_type.go (hunk 1 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/bucket_type.go]: fixed

File: action/protocol/staking/contractstaking/bucket_type.go (hunk 4 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/bucket_type.go]: fixed

File: action/protocol/staking/contractstaking/bucket_info.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/bucket_info.go]: fixed

**Pull request merged**